ESQL: ST_EXTENT_AGG optimize envelope extraction from doc-values for cartesian_shape

GalLalouche opened the pull request:
This PR optimizes `ST_EXTENT_AGG` when performed on Cartesian shapes.

When we cartesian shapes to Lucene, we encode additional information in the binary, including the extent, so we can read the extent directly from the binary encoding instead of (re-)computing it per shape, and replace the (possibly very complicated) shape with a rectangle. At the moment, this is only done for Cartesian shapes, since for Geo shapes, we need to take dateline wrapping into account, which means it can't be directly encoded as a rectangle. We will deal with geo shapes in a future PR.

GalLalouche committed:
ESQL: ST_EXTENT_AGG binary extent optimization

GalLalouche committed:
ESQL: ST_EXTENT_AGG binary extent optimization

GalLalouche committed:
ESQL: ST_EXTENT_AGG binary extent optimization

GalLalouche committed:
Update docs/changelog/118802.yaml

GalLalouche committed:
Update docs/changelog/118802.yaml

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

elasticsearchmachine commented:
Hi @GalLalouche, I've created a changelog YAML for you.

GalLalouche committed:
Merge branch 'optimization/st_extent' of github.com:GalLalouche/elasticsearch into optimization/st_extent

GalLalouche committed:
Manually modify changelog

GalLalouche committed:
Manually modify changelog

GalLalouche committed:
Fix TODO

GalLalouche committed:
Fix TODO

GalLalouche committed:
Small refactors

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into optimization/st_extent

craigtaverner committed:
Better changelog message

craigtaverner commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -6912,12 +7066,23 @@ private EsQueryExec assertChildIsGeoPointExtract(UnaryExec parent, boolean useDo
     }
 
     private EsQueryExec assertChildIsExtractedAsDocValues(UnaryExec parent, boolean useDocValues, DataType dataType) {
+        // TODO(gal) why is this OK To vacuously true?
```
What does this question mean? The line after it asserts that the child node is a FieldExtractExec, and that assertion is reflected in the method name above too.

craigtaverner commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -6912,12 +7066,23 @@ private EsQueryExec assertChildIsGeoPointExtract(UnaryExec parent, boolean useDo
+        return source(extract.child());
+    }
+
+    private static EsQueryExec assertChildIsExtractedAsBounds(UnaryExec parent, DataType dataType) {
```
I presume we could have combined these two methods, and replace the boolean useDocValues with a FieldExtractPreference, which would just switch which collection to test.

GalLalouche replied:
Pretty large refactoring, but I've replaced all the `Boolean`s with the enum 😎.

craigtaverner committed:
Reverted remove line

craigtaverner reviewed (changes requested):
I think there are still a few things to fix.

craigtaverner commented on server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:
```diff
@@ -676,11 +676,27 @@ public enum FieldExtractPreference {
+            return columnReader ? DOC_VALUES : NONE;
+        }
+
+        public boolean isColumnReader() {
```
Later I did see some code that could have used this, but at that point I also wondered if it would be better to just stick with the original boolean.

GalLalouche commented on server/src/main/java/org/elasticsearch/index/mapper/AbstractShapeGeometryFieldMapper.java:
```diff
@@ -69,6 +78,79 @@ protected Object nullValueAsSource(T nullValue) {
+                        var extent = reader.getExtent();
+                        // This is rather silly: an extent is already encoded as ints, but we convert it to Rectangle to
+                        // preserve its properties as a WKB shape, only to convert it back to ints when we compute the
+                        // aggregation. An obvious optimization would be to avoid this back-and-forth conversion.
```
I know, I'm writing it down to signify that this is not an oversight :)

GalLalouche committed:
Merge branch 'optimization/st_extent' of github.com:GalLalouche/elasticsearch into optimization/st_extent

craigtaverner committed:
Merge branch 'main' into optimization/st_extent

craigtaverner reviewed (changes requested):
There were no tests that cover the case where shapes were ingested without doc-values. If these existed, they would have failed since the current optimization triggers even if there are no doc-values. There are a few things to do to cover this scenario:
* Update PhysicalPlanOptimizerTests to have the airportCityBoundary index with and without doc-values and make assertions that the plan is only optimized to load extents from doc values if they are there.
* Add csv-spec tests that have cartesian_shapes with and without doc-values, and assert that the queries run with the same results in both.
* Add SpatialPushDownCartesianShapeIT tests that test the same results with and without doc-values and filter pushdown

For this PR it is sufficient to have either the csv-spec, or the SpatialPushDownCartesianShapeIT tests asserting the same results. Having both is even better.

GalLalouche commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -2965,6 +2974,151 @@ public void testSpatialTypesAndStatsExtentAndCentroidUseDocValues() {
+        // below the exchange (in data node) the aggregation is field optimized.
+        assertAggregation(agg, "extent", SpatialExtent.class, GEO_SHAPE, withDocValues);
+        var fieldExtractExec = as(agg.child(), FieldExtractExec.class);
+        assertThat(fieldExtractExec.boundAttributes().stream().map(a -> a.sourceText()).toList(), equalTo(List.of("city_boundary")));
```
There was no CSV test to check extent on non-doc-values Cartesian shapes, so I added one.

GalLalouche commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -2965,6 +2974,151 @@ public void testSpatialTypesAndStatsExtentAndCentroidUseDocValues() {
+     *       \_AggregateExec[[],[SPATIALSTEXTENT(location{f}#48,true[BOOLEAN]) AS extent],INITIAL,[
+     * minNegX{r}#73, minPosX{r}#74, maxNegX{rb#75, maxPosX{r}#76, maxY{r}#77, minY{r}#78],21]
+     *         \_FieldExtractExec[location{f}#48][location{f}#48]
+     *           \_EsQueryExec[airports], indexMode[standard], query[{"exists":{"field":"location","boost":1.0}}][
```
Nice catch!

GalLalouche committed:
Review fixes

GalLalouche committed:
Merge branch 'main' into optimization/st_extent

GalLalouche commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -2965,6 +2974,151 @@ public void testSpatialTypesAndStatsExtentAndCentroidUseDocValues() {
+     */
+    public void testSpatialTypesAndStatsExtentOfShapesUsesBinaryExtraction() {
+        for (String query : new String[] { "from airports_city_boundaries | stats extent = st_extent_agg(city_boundary)", }) {
+            var withDocValues = false;
```
Removed.

craigtaverner committed:
Merge branch 'main' into optimization/st_extent

craigtaverner commented on test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java:
```diff
@@ -1484,16 +1483,20 @@ private void testBlockLoader(boolean syntheticSource, boolean columnReader) thro
         SourceLoader sourceLoader
     ) throws IOException {
-        BlockLoader loader = blockReaderSupport.getBlockLoader(columnReader);
+        var columnReader = switch (fieldExtractPreference) {
```
Not really. My comment here was about continuing to use fieldExtractPreference, which we agreed on.

craigtaverner replied:
However, I don't understand why `EXTRACT_SPATIAL_BOUNDS` has columnReader=false? It should be true, because we'll be reading doc-values, which is column oriented.

GalLalouche replied:
Well, it's mostly to avoid triggering the old behavior. Ideally nothing would use that Boolean, but since some old code *does* use it, I want this to take a different path.

GalLalouche replied:
OK, upon re-reviewing the code, I see now what you mean. I've changed it to `EXTRACT_SPATIAL_BOUNDS` is column reader.

craigtaverner replied:
Perhaps this code is simply never tested, which is why we see no errors here?

GalLalouche replied:
Yes, nothing is exercising the path of `EXTRACT_SPATIAL_BOUNDS`. I think I was trying to use it before I gave up on it and just tested `AbstractShapeGeometryFieldMapper` directly! Let me revert this whole thing back to Boolean. No reason to have stamp coupling :)

craigtaverner commented on server/src/test/java/org/elasticsearch/index/mapper/AbstractShapeGeometryFieldMapperTests.java:
```diff
@@ -0,0 +1,92 @@
+
+    private void testBoundsBlockLoaderAux(
+        CoordinateEncoder encoder,
+        Supplier<Geometry> generator,
```
Nope. Still unused.

craigtaverner commented on server/src/test/java/org/elasticsearch/index/mapper/AbstractShapeGeometryFieldMapperTests.java:
```diff
@@ -0,0 +1,92 @@
+        Function<String, ShapeIndexer> indexerFactory,
+        Function<Geometry, Optional<Rectangle>> visitor
+    ) throws IOException {
+        var geometries = IntStream.range(0, 20).mapToObj(i -> ShapeTestUtils.randomGeometryWithoutCircle(0, false)).toList();
```
It still does not call generator.

craigtaverner committed:
Merge branch 'optimization/st_extent' of github.com:GalLalouche/elasticsearch into optimization/st_extent

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into optimization/st_extent

GalLalouche committed:
Review fixes

craigtaverner commented on server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:
```diff
@@ -677,10 +677,16 @@ public enum FieldExtractPreference {
+        NONE;
+
+        public static FieldExtractPreference forColumnReader(boolean columnReader) {
+            return columnReader ? DOC_VALUES : NONE;
```
This does not really make sense to me. Both DOC_VALUES and EXTRACT_SPATIAL_BOUNDS will require a column reader. I suspect the situation is that the places this is used actually never test the EXTRACT_SPATIAL_BOUNDS case, so this does not matter. But once we add tests for that, it will.

GalLalouche replied:
Yes, this is for backward compatibility. I'll just inline it so it's clearer.

GalLalouche committed:
Fix test docs, revert MapperTestCase changes

GalLalouche committed:
Merge branch 'optimization/st_extent' of github.com:GalLalouche/elasticsearch into optimization/st_extent

GalLalouche commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -2965,6 +2974,151 @@ public void testSpatialTypesAndStatsExtentAndCentroidUseDocValues() {
+     *     \_ExchangeExec[[minNegX{r}#52, minPosX{r}#53, maxNegX{r}#54, maxPosX{r}#55, maxY{r}#56, minY{r}#57],true]
+     *       \_FragmentExec[filter=null, estimatedRowSize=0, reducer=[], fragment=[
+     * Aggregate[STANDARD,[],[SPATIALSTEXTENT(location{f}#48,true[BOOLEAN]) AS extent]]
+     *         \_EsRelation[airports][abbrev{f}#44, city{f}#50, city_location{f}#51, coun..]]]
```
Fixed (I hope).

GalLalouche committed:
Inline forColumnReader

craigtaverner committed:
Added missing test for bounds extraction from cartesian_shape doc-values

And made a few updates to comments and javadocs

craigtaverner committed:
Merge branch 'optimization/st_extent' of github.com:GalLalouche/elasticsearch into optimization/st_extent

craigtaverner committed:
Merge branch 'main' into optimization/st_extent

craigtaverner committed:
Fixed the test that was testing the wrong thing

craigtaverner committed:
Merge branch 'main' into optimization/st_extent

craigtaverner committed:
Added bounds extracting with grouping tests and fixed bug with multiple extractions

When there are multiple FieldExtractExec nodes, it is incorrect to allow the modification of the foundAttributes, but should take a copy, and modify the copy.

craigtaverner reviewed (approved):
With the latest tests and the grouping bug fixed, I think this looks good to go!

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/119187)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/119187">](https://github.com/elastic/elasticsearch/pull/119187) |
